### PR TITLE
Comprehend Moderation 0.1

### PR DESCRIPTION
**Pull request by nikhilkjha**
This PR streamlines the #9665 configuration object and logic to enable chain moderation via the Amazon Comprehend service, allowing you to detect and redact PII, Toxic, and Intent information in the LLM prompt or answer. 
This implementation simplifies the configuration and types it to specific class/model.

### Usage sample

```python
from langchain_experimental.comprehend_moderation import (BaseModerationConfig, 
                                 ModerationIntentConfig, 
                                 ModerationPiiConfig, 
                                 ModerationToxicityConfig
)

pii_config = ModerationPiiConfig(
    labels=["SSN"],
    redact=True,
    mask_character="X"
)

toxicity_config = ModerationToxicityConfig(
    threshold=0.5
)

intent_config = ModerationIntentConfig(
    threshold=0.5
)

moderation_config = BaseModerationConfig(
    filters=[pii_config, toxicity_config, intent_config]
)

comp_moderation_with_config = AmazonComprehendModerationChain(
    moderation_config=moderation_config, #specify the configuration
    client=comprehend_client,            #optionally pass the Boto3 Client
    verbose=True
)

template = """Question: {question}

Answer:"""

prompt = PromptTemplate(template=template, input_variables=["question"])

responses = [
    "Final Answer: A credit card number looks like 1289-2321-1123-2387. A fake SSN number looks like 323-22-9980. John Doe's phone number is (999)253-9876.", 
    "Final Answer: This is a really shitty way of constructing a birdhouse. This is fucking insane to think that any birds would actually create their motherfucking nests here."
]
llm = FakeListLLM(responses=responses)

llm_chain = LLMChain(prompt=prompt, llm=llm)

chain = ( 
    prompt 
    | comp_moderation_with_config 
    | {llm_chain.input_keys[0]: lambda x: x['output'] }  
    | llm_chain 
    | { "input": lambda x: x['text'] } 
    | comp_moderation_with_config 
)

try:
    response = chain.invoke({"question": "A sample SSN number looks like this 123-456-7890. Can you give me some more samples?"})
except Exception as e:
    print(str(e))
else:
    print(response['output'])

```

### Output

```python
> Entering new AmazonComprehendModerationChain chain...
Running AmazonComprehendModerationChain...
Running pii Validation...
Running toxicity Validation...
Running intent Validation...

> Finished chain.


> Entering new AmazonComprehendModerationChain chain...
Running AmazonComprehendModerationChain...
Running pii Validation...
Running toxicity Validation...
Running intent Validation...

> Finished chain.
Final Answer: A credit card number looks like 1289-2321-1123-2387. A fake SSN number looks like XXXXXXXXXXXX John Doe's phone number is (999)253-9876.
```

### CODE REVIEW

File: libs/experimental/langchain_experimental/comprehend_moderation/toxicity.py [hunk 1]
```diff
@@ -134,76 +132,34 @@ def validate(
...
+            toxicity_found = False
+            threshold = config.get("threshold")
+            toxicity_labels = config.get("labels")
+
+            if not toxicity_labels:
+                for item in response["ResultList"]:
+                    for label in item["Labels"]:
+                        if label["Score"] >= threshold:
+                            toxicity_found = True
+                            break
```
Comment (3coins) [inline libs/experimental/langchain_experimental/comprehend_moderation/toxicity.py:144]: Nit: Better to create a model to represent this, and move the nesting of this computation to those models. For example:

```python
class Label(BaseModel, extra=Extra.allow):
    Score: int

    def istoxic(self, threshold: int) -> bool:
        return self.Score >= threshold

class Result(BaseModel, extra=Extra.allow):
    Labels: Sequence[Label] = []
    
    def istoxic(self, threshold: int) -> bool:
        for label in Labels:
            return True if label.istoxic(threshold)

        return False

class Response(BaseModel, extra=Extra.allow):
    ResultList: Sequence[Result]
    
    def istoxic(self, threshold: int) -> bool:
        for result in ResultList:
            return True if result.istoxic(threshold)

        return False

```

Then the code becomes:

```python
_response = Response.parse_obj(response)
if not toxicity_labels:
    toxicity_found = _response.istoxic(threshold)
```

File: libs/experimental/langchain_experimental/comprehend_moderation/toxicity.py [hunk 2]
```diff
@@ -134,76 +132,34 @@ def validate(
...
-                                )
-                            )
-                        raise ModerationToxicityError
+                for item in response["ResultList"]:
+                    for label in item["Labels"]:
+                        if (
+                            label["Name"] in toxicity_labels
+                            and label["Score"] >= threshold
+                        ):
+                            toxicity_found = True
```
Comment (3coins) [inline libs/experimental/langchain_experimental/comprehend_moderation/toxicity.py:152]: See the comment above.

**Comment by anjanvb** [comment]
Copy that @3coins, will make those changes.

File: libs/experimental/langchain_experimental/comprehend_moderation/amazon_comprehend_moderation.py [hunk 3]
```diff
@@ -21,8 +24,11 @@ class AmazonComprehendModerationChain(Chain):
     input_key: str = "input"  #: :meta private:
     """Key used to fetch/store the input in data containers. Defaults to `input`"""
 
-    moderation_config: Optional[Dict[str, Any]] = None
-    """Configuration settings for moderation"""
+    moderation_config: Optional[BaseModerationConfig] = BaseModerationConfig()
```
[the remark this answers is not in the document]
Comment (iamwrick) [reply libs/experimental/langchain_experimental/comprehend_moderation/amazon_comprehend_moderation.py:27]: fixed

File: libs/experimental/langchain_experimental/comprehend_moderation/amazon_comprehend_moderation.py [hunk 4]
```diff
@@ -21,8 +24,11 @@ class AmazonComprehendModerationChain(Chain):
...
+    """
+    Configuration settings for moderation, 
+    defaults to BaseModerationConfig with default values
+    """
 
     client: Optional[Any]
```
[the remark this answers is not in the document]
Comment (iamwrick) [reply libs/experimental/langchain_experimental/comprehend_moderation/amazon_comprehend_moderation.py:33]: resolved

File: libs/experimental/langchain_experimental/comprehend_moderation/base_moderation_config.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+from typing import Any, List, Optional
+
+from pydantic import BaseModel
+
+
+class ModerationPiiConfig(BaseModel):
+    """Threshold for PII confidence score, defaults to 0.5 i.e. 50%"""
```
[the remark this answers is not in the document]
Comment (iamwrick) [reply libs/experimental/langchain_experimental/comprehend_moderation/base_moderation_config.py:7]: fixed

File: libs/experimental/langchain_experimental/comprehend_moderation/base_moderation_config.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    threshold: Optional[float] = 0.5
```
[the remark this answers is not in the document]
Comment (iamwrick) [reply libs/experimental/langchain_experimental/comprehend_moderation/base_moderation_config.py:9]: fixed

File: libs/experimental/langchain_experimental/comprehend_moderation/intent.py [hunk 7]
```diff
@@ -30,20 +29,17 @@ def _get_arn(self) -> str:
         intent_endpoint = "document-classifier-endpoint/prompt-intent"
         return f"arn:aws:{service}:{region_name}:aws:{intent_endpoint}"
 
-    def validate(
-        self, prompt_value: str, config: Optional[Dict[str, Any]] = None
-    ) -> str:
+    def validate(self, prompt_value: str, config: Any = None) -> str:
```
[the remark this answers is not in the document]
Comment (iamwrick) [reply libs/experimental/langchain_experimental/comprehend_moderation/intent.py:32]: Will trace it in our next release and make changes accordingly

File: libs/experimental/langchain_experimental/comprehend_moderation/pii.py [hunk 8]
```diff
@@ -23,33 +23,19 @@ def __init__(
         self.callback = callback
         self.unique_id = unique_id
 
-    def validate(
-        self, prompt_value: str, config: Optional[Dict[str, Any]] = None
-    ) -> str:
-        from langchain_experimental.comprehend_moderation.base_moderation_enums import (
-            BaseModerationActions,
+    def validate(self, prompt_value: str, config: Any = None) -> str:
```
[the remark this answers is not in the document]
Comment (iamwrick) [reply libs/experimental/langchain_experimental/comprehend_moderation/pii.py:26]: Will trace it in our next release and make changes accordingly

File: libs/experimental/langchain_experimental/comprehend_moderation/pii.py [hunk 9]
```diff
@@ -23,33 +23,19 @@ def __init__(
...
+        redact = config.get("redact")
+        return (
+            self._contains_pii(prompt_value=prompt_value, config=config)
+            if not redact
+            else self._detect_pii(prompt_value=prompt_value, config=config)
```
[the remark this answers is not in the document]
Comment (iamwrick) [reply libs/experimental/langchain_experimental/comprehend_moderation/pii.py:31]: fixed

File: libs/experimental/langchain_experimental/comprehend_moderation/pii.py [hunk 10]
```diff
@@ -23,33 +23,19 @@ def __init__(
...
-                if action == BaseModerationActions.STOP
-                else self._detect_pii(prompt_value=prompt_value, config=config)
-            )
-        else:
-            return self._contains_pii(prompt_value=prompt_value)
-
-    def _contains_pii(
-        self, prompt_value: str, config: Optional[Dict[str, Any]] = None
-    ) -> str:
+    def _contains_pii(self, prompt_value: str, config: Any = None) -> str:
```
[the remark this answers is not in the document]
Comment (iamwrick) [reply libs/experimental/langchain_experimental/comprehend_moderation/pii.py:34]: Will trace it in our next release and make changes accordingly

File: libs/experimental/langchain_experimental/comprehend_moderation/toxicity.py [hunk 11]
```diff
@@ -103,16 +105,12 @@ def _split_paragraph(
         # Add any remaining sentences
         if current_chunk:
             chunks.append(current_chunk)
-
         return chunks
 
-    def validate(
-        self, prompt_value: str, config: Optional[Dict[str, Any]] = None
-    ) -> str:
+    def validate(self, prompt_value: str, config: Any = None) -> str:
```
[the remark this answers is not in the document]
Comment (iamwrick) [reply libs/experimental/langchain_experimental/comprehend_moderation/toxicity.py:110]: Will trace it in our next release and make changes accordingly

File: libs/experimental/langchain_experimental/comprehend_moderation/toxicity.py (hunk 1 above)
Comment (iamwrick) [reply libs/experimental/langchain_experimental/comprehend_moderation/toxicity.py:144]: Will trace it in our next release and make changes accordingly

File: libs/experimental/langchain_experimental/comprehend_moderation/toxicity.py (hunk 2 above)
Comment (iamwrick) [reply libs/experimental/langchain_experimental/comprehend_moderation/toxicity.py:152]: Will trace it in our next release and make changes accordingly

**Comment by 3coins** [comment]
@nikhilkjha 
Please close this PR in favor of #10104.

**Comment by nikhilkjha** [comment]
Closing this out as the changes are on #10104

**Pull request closed without merging**